src/adapter/breakpoints/logPoint.ts

```diff
@@ -0,0 +1,118 @@
+/*---------------------------------------------------------
+ * Copyright (C) Microsoft Corporation. All rights reserved.
+ *--------------------------------------------------------*/
+
+import * as ts from 'typescript';
+import { assert, logger } from '../../common/logging/logger';
+import { invalidLogPointSyntax } from '../../dap/errors';
+import { LogTag } from '../../common/logging';
+/**
+ * Statments which we should not prefix with `return {}`
+ */
+const unreturnable: ReadonlySet<ts.SyntaxKind> = new Set([
+  ts.SyntaxKind.ReturnStatement,
+  ts.SyntaxKind.IfStatement,
+  ts.SyntaxKind.DoStatement,
+  ts.SyntaxKind.WhileStatement,
+  ts.SyntaxKind.ForStatement,
+  ts.SyntaxKind.ForInStatement,
+  ts.SyntaxKind.ForOfStatement,
+  ts.SyntaxKind.ContinueStatement,
+  ts.SyntaxKind.BreakStatement,
+  ts.SyntaxKind.ReturnStatement,
+  ts.SyntaxKind.WithStatement,
+  ts.SyntaxKind.SwitchStatement,
+  ts.SyntaxKind.LabeledStatement,
+  ts.SyntaxKind.TryStatement,
+  ts.SyntaxKind.ThrowStatement,
+  ts.SyntaxKind.DebuggerStatement,
+  ts.SyntaxKind.VariableDeclaration,
+  ts.SyntaxKind.VariableDeclarationList,
+]);
```

Comment:
The problem here is that if the breakpoint just has side-effect code like `while (foo) { bar(); }`, emitting `return while (foo)` is invalid. Instead of having a denylist of grammars, we could just `try { new Function('return ${stmt}') }` to see if it's valid. Not quite as fast or 'pure', but probably more correct. Could go either way, thoughts?

Reply:
I would probably go with the dynamic approach. At least it would be future-proof. But I could go either way.